tests/cycle_maybe_changed_after.rs

```diff
@@ -0,0 +1,213 @@
+//! Tests for incremental validation for queries involved in a cycle.
+mod common;
+
+use crate::common::EventLoggerDatabase;
+use salsa::{CycleRecoveryAction, Database, Durability, Setter};
+
+#[salsa::input(debug)]
+struct Input {
+    value: u32,
+    max: u32,
+}
+
+#[salsa::interned(debug)]
+struct Output<'db> {
+    #[return_ref]
+    value: u32,
+}
+
+#[salsa::tracked(cycle_fn=query_a_recover, cycle_initial=query_a_initial)]
+fn query_c<'db>(db: &'db dyn salsa::Database, input: Input) -> u32 {
+    query_d(db, input)
+}
+
+#[salsa::tracked]
+fn query_d<'db>(db: &'db dyn salsa::Database, input: Input) -> u32 {
+    let value = query_c(db, input);
+    if value < input.max(db) * 2 {
+        // Only the first iteration depends on value but the entire
+        // cycle must re-run if input changes.
+        let result = value + input.value(db);
+        Output::new(db, result);
+        result
+    } else {
+        value
+    }
+}
+
+fn query_a_initial(_db: &dyn Database, _input: Input) -> u32 {
+    0
+}
+
+fn query_a_recover(
```

Comment:
Maybe these shouldn't be named `query_a_*` since they are used by both `query_a` and `query_d`? Looks a bit weird above.